### PR TITLE
Add non-hack version LoRA: lora module and inject function, lora finetuning and inference, and unit test

### PR DESCRIPTION
**Pull request by SamitHuang**
- Motivation
Add features of LoRA finetuning and inference.

- TODO
 1. Save LoRA trainable params only in a smaller checkpoint. Load pretrained ckpt then lora finetuned ckpt.
 2. Improve fine-tuning results. e.g. improve training recipe and tune hyper-param for training. Need to implement gradient clip feature. 
 3. Display generation result and eval result.

### CODE REVIEW

**Comment by congw729** [comment]
LGTM

**Review by congw729 — APPROVED** [review]
LGTM
